Add QR factorization by exposing LAPACK functions XGEQRF, XORMQR, XUNMQR

gau27 opened the pull request:
Fixes #440 : 

Ruby Methods Added : (Requires nmatrix-lapacke plugin)

**geqrf!**  :  Ruby method to interface with LAPACKE_XGEQRF
**ormqr** :  Ruby method to interface with LAPACKE_XORMQR (Necessary for generating Q)
**unmqr** : Ruby method to interface with LAPACKE_XUNMQR (Complex type equivalent of ormqr)
**factorize_qr** : Friendly user interface to obtain QR decomposition.
 
Example Usage : 

```ruby
[1] pry(main)> a = N.new([3,3], [12.0, -51, 4, 6, 167, -68, -4, 24, -41])
=> 
[
  [12.0, -51.0,   4.0]   [ 6.0, 167.0, -68.0]   [-4.0,  24.0, -41.0] ]

[3] pry(main)> q,r = a.factorize_qr
=> [
[
   [-0.8571428571428572,  0.3942857142857143,  0.33142857142857146]
   [-0.4285714285714286, -0.9028571428571428, -0.03428571428571425]
   [0.28571428571428575, -0.1714285714285714,   0.9428571428571428]
 ],
 
[
   [-14.0,               -21.0, 14.000000000000002]
   [  0.0, -175.00000000000003,  70.00000000000001]
   [  0.0,                 0.0,              -35.0]
 ]]

[4] pry(main)> q.dot(r)
=> 
[
  [              12.0, -51.000000000000014,  4.000000000000002]
  [               6.0,  167.00000000000003, -68.00000000000001]
  [-4.000000000000001,                24.0,              -41.0]
]
```
20 LAPACKE Source Files Added : 
```ruby
/ext/nmatrix_lapacke/lapacke/src/

  - 4 x  XGEQRF 
  - 4 x  XGEQRF_WORK
  - 2 x XORMQR
  - 2 x XORMQR_WORK
  - 2 x UNMQR
  - 2 x UNRMQR_WORK

/ext/nmatrix_lapacke/lapacke/utils/
 
 - lapacke_c_nancheck.c
 - lapacke_z_nancheck.c
 - lapacke_d_nancheck.c
 - lapacke_s_nancheck.c
```

cc @wlevine : Please have a look

gau27 commented:
I tried to use row major input to avoid the overhead of transposing but for random tests I would get a segfault on quitting pry, and accuracy was off for non-square matrices.

https://gist.github.com/gau27/d3081b5f3a1d8157bccb

wlevine commented on ext/nmatrix/math/util.h:
```diff
@@ -145,4 +145,24 @@ static inline char lapack_evd_job_sym(VALUE op) {
+ * Related to obtaining Q in QR factorizationa after calling lapack_geqrf
+ */
+
+static inline char lapack_qr_side(VALUE op) {
```
The symbols accepted by this function should probably match those used `blas_side_sym` above. Also, rename this to `lapacke_side_sym`

wlevine commented on ext/nmatrix/math/util.h:
```diff
@@ -145,4 +145,24 @@ static inline char lapack_evd_job_sym(VALUE op) {
+  return 'L';
+}
+
+static inline char lapack_qr_transpose(VALUE op) {
```
You can just use the existing `lapacke_transpose_sym`

wlevine commented:
Nice work, @gau27. I'll try to finish looking at the code tonight.

Do you think it's more useful to return the full Q and R matrices or the reduced ones? What do other libraries do?

[this wikipedia page](https://en.wikipedia.org/wiki/QR_decomposition#Using_for_solution_to_linear_inverse_problems) describes how to use the QR factorization for least squares: we want the product R_1^(-1)(Q_1^T b). If we wanted to do this would we need to all three steps separately: use geqrf to get Q and R, then ormqr to get (Q_1^T b), and then trsm to solve R_1 x = Q_1^T b? Or is there another LAPACK function which will combine some of these steps? @agisga, does this sound right to you? @gau27, I'm not saying you should include this in this pull request, I'm just trying to figure out what the next step is.

wlevine commented on lib/nmatrix/math.rb:
```diff
@@ -155,6 +155,57 @@ def getrf!
+    
+    temp = self.transpose
+    tau = NMatrix.new([self.shape.min,1], dtype: self.dtype)
+    NMatrix::LAPACK::lapacke_geqrf(:col, self.shape[0], self.shape[1], temp, self.shape[0], tau)
```
When I try this it works fine with `:row`. For row-major you should be using `shape[1]` for `lda`

wlevine replied:
Nothing in math.rb should directly call a lapacke_* function. All calls to lapacke_* functions should be in lapacke.rb. You can just move `#geqrf!` directly to lapacke.rb, I think.

wlevine commented on lib/nmatrix/math.rb:
```diff
@@ -155,6 +155,57 @@ def getrf!
+  # The QR factorization is A = QR, where Q is orthogornal and R is Upper Triangular
+  # +A+ is overwritten with the elements of R and Q with Q being represented by the 
+  # elements below A's diagonal and an array of scalar factors in the output NMatrix. 
+  #  
```
Either expand this explanation or link to the documentation.

wlevine commented on lib/nmatrix/math.rb:
```diff
@@ -155,6 +155,57 @@ def getrf!
+    tau
+  end
+
+  def ormqr(tau, side=:left, transpose=:no)
```
Allow this to act on arbitrary matrices.

Also, add docs for this function.

wlevine commented on lib/nmatrix/math.rb:
```diff
@@ -232,6 +283,49 @@ def factorize_lu with_permutation_matrix=nil
+  def factorize_qr
+    raise(NotImplementedError, "only implemented for dense storage") unless self.stype == :dense
+    raise(ShapeError, "Input must be a 2-dimensional matrix to have a QR decomposition") unless self.dim == 2
+    raise(ShapeError, "Input must have shape[0] > shape[1]") unless self.shape[0] >= self.shape[1]
```
Why this requirement?

gau27 replied:
I noticed that mentioned as a constraint on most lecture notes / [Wikipedia](https://en.wikipedia.org/wiki/QR_decomposition#Rectangular_matrix). My understanding is that the result of geqrf on matrices with more columns than rows would not be a meaningful QR decomposition. I'll look at this after I'm done with the tests.

https://en.wikipedia.org/wiki/QR_decomposition#Rectangular_matrix

agisga replied:
Of course, in order to get a least squares solution to a linear system a matrix with more rows than columns is required, because otherwise we have an underdetermined system (i.e. more unknowns than equations).
But in general, QR factorization of a matrix with more columns than rows definitely makes sense. I personally have used it for dimensional reduction in some context (i.e. take a n x p matrix A with n << p, do QR=A, continue computation with much smaller n x n matrix Q).

gau27 replied:
@agisga : Thanks for pointing that out, I'll make the changes

agisga commented:
> use geqrf to get Q and R, then ormqr to get (Q_1^T b), and then trsm to solve R_1 x = Q_1^T b? Or is there another LAPACK function which will combine some of these steps? @agisga, does this sound right to you?

I think that's right, that's how I would try to do it too. At least, I cannot think of a better way to do it. I'm not familiar with other LAPACK functions though.

wlevine commented on ext/nmatrix_lapacke/lapacke/utils/lapacke_c_nancheck.c:
```diff
@@ -0,0 +1,51 @@
+/*****************************************************************************
```
It looks like you forgot lapacke_z_nancheck

gau27 replied:
Thanks for pointing that out, I've included it in the most recent commit.

wlevine commented on ext/nmatrix_lapacke/math_lapacke/lapacke_templates.h:
```diff
@@ -64,6 +64,85 @@ inline int lapacke_getrf(const enum CBLAS_ORDER order, const int m, const int n,
+template <typename DType>
+inline int unmqr(const enum CBLAS_ORDER order, char side, char trans, const int m, const int n, const int k, DType* a, const int lda, DType* tau, DType* c, const int ldc) {
+  //We don't want to call the internal implementation since the the CLAPACK interface is slightly different than the LAPACKE.
+  rb_raise(rb_eNotImpError, "lapacke_unmqr not implemented for non_BLAS dtypes.");
```
non-complex dtypes

gau27 commented:
@wlevine : I've made the following changes according to your suggestions :-

- `geqrf!` , `ormqr`, `unmqr` have been moved to `lapacke.rb`. Dummy functions and docs have been put in math.rb to produce a more meaningful error when LAPACKE isn't plugged in.

-  switched to :row format (I totally missed the LDA argument, thanks for pointing that out) 

-  `ormqr` /  `unmqr` have been modified to accept any arbitrary matrix as Q's multiplicand. (The default behaviour uses the identity matrix to produce Q)

-  added tests and completed docs / cleanup.

mohawkjohn commented:
@wlevine What do you think?

wlevine commented on spec/spec_helper.rb:
```diff
@@ -48,6 +48,29 @@
+Q_SOLUTION_ARRAY_3 = [-0.7724247413634004, -0.026670393594597247, -0.6345460653374136, 
+                      -0.5777485870360393,  -0.38541856437557026,  0.7194853024298236,
+                      -0.26375478973384403,   0.9223563413020934, 0.28229805268947933]
+
```
I would only store these in here if they are shared between different tests.

gau27 replied:
`factorize_qr` in `math_spec.rb` and the lapacke methods in 'lapacke_spec.rb' share these values for their test cases. Is that okay?

wlevine commented on lib/nmatrix/lapacke.rb:
```diff
@@ -232,4 +232,111 @@ def solve(b, opts = {})
+  #
+  # call-seq:
+  #     ormqr(tau) -> NMatrix
+  #     ormqr(tau, :left, false, C) -> NMAtrix
```
These argument names should match the ones below. Also NMAtrix -> NMatrix :)

wlevine commented on lib/nmatrix/lapacke.rb:
```diff
@@ -232,4 +232,111 @@ def solve(b, opts = {})
+  # * +tau+ - vector containing scalar factors of elementary reflectors
+  # * +side+ - direction of multiplication [:left, :right]
+  # * +transpose+ - Apply Q with or without transpose [false, :transpose] 
+  # * +multiplicand+ - Multiplicand of Q * C, no argument assumes C = I
```
Q*C or C*Q depending on the arguments. Also, you should mention that this argument will be modified.

wlevine commented:
@gau27 Thanks for making these changes. I like that you added a lot of specs to test a variety of situations, but still I would like to see a few more tests:  #geqrf! amd #ormqr should be tested directly (one test for each is fine), and maybe we should have at least one test of ormqr with non-default options (side or transpose). My other comments were very minor, so this is almost ready to merge.

gau27 commented:
@wlevine : Thank you for your patient reviews :) , I've rectified the docs and added some more tests for the following cases :-

```geqrf
ormqr(tau) 
ormqr(tau, :left, :transpose)
ormqr(tau, :right, false, c)
```

wlevine commented on lib/nmatrix/lapacke.rb:
```diff
@@ -232,4 +232,122 @@ def solve(b, opts = {})
+  #
+  # call-seq:
+  #     ormqr(tau) -> NMatrix
+  #     ormqr(tau, :right, :transpose, c) -> NMatrix
```
What I meant here is `ormqr(tau, side, transpose, c)`

gau27 replied:
Oops, fixed it.

wlevine commented on lib/nmatrix/lapacke.rb:
```diff
@@ -232,4 +232,122 @@ def solve(b, opts = {})
+  def unmqr(tau, side=:left, transpose=false, c=nil)
+    raise(StorageTypeError, "ATLAS functions only work on dense matrices") unless self.dense?
+    raise(TypeError, "Works only on complex matrices, use ormqr for normal floating point matrices") unless self.complex_dtype?
+    raise(TypeError, "c must have the same dtype as the calling NMatrix") if c and c.dtype != self.dtype
```
This is a good check to add, but I think it also applies to ormqr, no?

gau27 replied:
Oh yes, in fact i had written it for ormqr first so I think I Ctrl-X'd it by mistake. Fixed.

gau27 commented:
Some specs are [failing](https://travis-ci.org/SciRuby/nmatrix/builds/116492007), in particular mostly Open Blas and that too some sign mismatches :-

```ruby
4) NMatrix::LAPACK functions implemented with LAPACKE interface calculates the multiplication c * Q using ormqr/unmqr after getrf!

     Failure/Error: expect(c).to be_within(err).of(x)

       expected [-3.8000000000000007, -1.6, -3.4000000000000012, 1.2000000000000004, -9.8, -0.7999999999999998, -6.4, 0.6000000000000001] to be within 1.0e-14 of [-3.8, 1.6, 3.4, 1.2, -9.8, 0.8, 6.4, 0.6]

     # ./spec/plugins/lapacke/lapacke_spec.rb:485:in `block (3 levels) in <top (required)>'

```
I've added another test to verify if the same test case can produce QR = A with the same test case. If that passes then I believe it's just an internal implementation difference that produces two valid factorizations with minor differences. I've already verified that the different Qs are indeed orthogonal but I don't have an OpenBLAS implementation on my local system to check any further.

gau27 commented:
I've used a non-singular square matrix in the most recent commit because it's expected to have a unique QR decomposition.

gau27 commented:
The builds are passing after changing the inputs for some tests, and the input that was failing the previous builds doesn't fail a test like `q.dot(r) == a` , so I think it's just a subtle difference in implementation of the intermediate outputs.

wlevine commented:
@mohawkjohn Looks ready to merge

mohawkjohn commented:
Thanks, @wlevine and @gau27!
